### PR TITLE
Update PowerShell extension to v2021.2.2

### PR DESCRIPTION
**Pull request by rjmholt**
This PR fixes https://github.com/PowerShell/vscode-powershell/issues/2993.

You'll need to upload [the VSIX asset](https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-vscode/vsextensions/PowerShell/2021.2.2/vspackage) to your own blob storage

### CODE REVIEW

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]
@rjmholt Unfortunately it looks like this version of the extension requires VS Code engine >= 1.52, we're only up to 1.51 in main and so we can't update this as it is.

Do you know what features in 1.52 you're using? We may be able to work around this if those features aren't something that ADS would use (such as debugging)

**Comment by rjmholt** [comment]
I believe we use it for the command explorer (which is a pane that shows currently defined commands in the UI): https://github.com/PowerShell/vscode-powershell/commit/cc74a42f5bc98d1f6f739e1ec1dd027401bbd3db

**Comment by Charles-Gagnon** [comment]
Shoot, well unfortunately the vscode-language-client was updated as well - and that's hardcoded to check the VS Code engine version on startup.

```
2/26/2021 1:50:24 PM [NORMAL] - Connecting to language service on pipe \\.\pipe\PSES_sxpedsvd.bwy...
2/26/2021 1:50:24 PM [ERROR] - The language service could not be started: 
2/26/2021 1:50:24 PM [ERROR] - Error: The language client requires VS Code version ^1.52.0 but received version 1.48.0
```

So there's pretty much nothing we can do until ADS is updated to VS Code engine >= 1.52 if you wanted to get this version.

Is there an earlier version that we could at least update to in the interim until we catch up to VS Code here?

**Comment by andschwa** [comment]
I believe you have the only earlier version, 2020.6.0. Since that version, as you identified we updated the `vscode-language-client` dependency from a prerelease at [7.0.0-next.14](https://github.com/microsoft/vscode-languageserver-node/blob/release/client%2F7.0.0-next.14/client/package.json) which required vscode 1.51, to their stable release at [7.0.0](https://github.com/microsoft/vscode-languageserver-node/blob/release/client%2F7.0.0/client/package.json) which came with a bump to 1.52. I wouldn't recommend we downgrade to a prerelease version, especially since that update was done in order to fix multiple [build and audit issues](https://github.com/PowerShell/vscode-powershell/pull/3099). I suggest sticking with the previous stable version until ADS updates to 1.52 (which must be around the corner 😄).

**Comment by Charles-Gagnon** [comment]
Sounds good, we're working on the latest VS Code merge now which should get us up to that point so hopefully that'll make it in for our March release.

**Comment by andschwa** [comment]
@Charles-Gagnon Just curious, any luck getting the VS Code update merged?

**Comment by andschwa** [comment]
No problem. There was a major issue with the 1.54.0 release anyway, but 1.54.2 is dropping today which resolves it.

**Comment by andschwa** [comment]
Hey @Charles-Gagnon, I was just wondering where you all were at with this? I'd also like to consider merging https://github.com/PowerShell/vscode-powershell/pull/3179 which would bring our minimum support VS Code to 1.53 (though I'd go higher if possible) so I want to know what minimum VS Code version you all need us to support. The current latest stable version of VS Code is 1.55.2. Thanks!

**Comment by Charles-Gagnon** [comment]
We're finishing up the VS Code merge to bring us up to 1.53 so once that's done we should be able to bring in the latest version.

https://github.com/microsoft/azuredatastudio/pull/14883

Beyond that though we haven't come to a decision on timeline for the other merges so can't say when we'd get 1.54+.

**Comment by andschwa** [comment]
Great news, thanks! We may even have a new stable extension for you by then 😄

**Comment by Charles-Gagnon** [comment]
When do you think you'll have the new version out? We just merged in the VS Code merge so we can get the insiders gallery updated tomorrow if it's going to be at least a few days.

**Comment by andschwa** [comment]
Ah, it'll be next few weeks, not days.

**Comment by Charles-Gagnon** [comment]
Sounds good. Could you : 

1. Merge in the latest from `release/extensions`
2. Remove the changes to extensionsGallery.json (this still won't work on stable, we'll update that when we do our next release)
2. Update the min azdata engine version to 1.29.0

Once that's done then we can do some testing tomorrow with the linked build and if things look good we'll get this one out in the wild while we wait on further updates.

**Comment by rjmholt** [comment]
@Charles-Gagnon this should be done now

**Comment by andschwa** [comment]
Thanks @rjmholt!

File: extensionsGallery-insider.json [hunk 1]
```diff
@@ -1450,7 +1450,7 @@
 								},
 								{
 									"key": "Microsoft.AzDataEngine",
-									"value": "*"
+									"value": "1.29.0"
```
Comment (Charles-Gagnon) [inline extensionsGallery-insider.json:1453]: This should be `>=1.29.0` (see the error in the check : https://github.com/microsoft/azuredatastudio/pull/14467/checks?check_run_id=2453507130)

Comment (rjmholt) [inline extensionsGallery-insider.json:1453]:
```suggestion
									"value": ">=1.29.0"
```

Comment (rjmholt) [reply extensionsGallery-insider.json]: This should be done now

**Comment by Charles-Gagnon** [comment]
Unfortunately we had an issue with the merge that's blocking us from releasing so we don't have a new insiders build today...I'll update once we get that resolved 😄

**Comment by Charles-Gagnon** [comment]
New insiders build is out! I'll install the extension and run through a few things myself, but would you mind doing some basic testing yourself to make sure things work as you expect given your familiarity with the feature? Would be good to make sure things look good before releasing it to the public (since rolling back isn't easy)

**Comment by Charles-Gagnon** [comment]
Here's the uploaded VSIX to try to make sure that the one we have uploaded is the correct version : https://sqlopsextensions.blob.core.windows.net/extensions/powershell/ms-vscode.PowerShell-2021.2.2.vsix

**Comment by Charles-Gagnon** [comment]
So I can install it fine and some features seem to work, but the cmdlet view is empty and in the logs I see a bunch of errors like this which seem to be coming from the extension. These seem familiar to you?

```
rejected promise not handled within 1 second: Error: Internal Error - System.Management.Automation.ParseException: At C:\Users\chgagnon\AppData\Local\Yarn\bin\tsc.ps1:5 char:13
+     *CYGWIN*) basedir=`cygpath -w "$basedir"`;;
+             ~Unexpected token ')' in expression or statement.

At C:\Users\chgagnon\AppData\Local\Yarn\bin\tsc.ps1:8 char:3
+ if [ -x "$basedir/pwsh" ]; then
+   ~Missing '(' after 'if' in if statement.

At C:\Users\chgagnon\AppData\Local\Yarn\bin\tsc.ps1:8 char:5
+ if [ -x "$basedir/pwsh" ]; then
+     ~Missing type name after '['.

At C:\Users\chgagnon\AppData\Local\Yarn\bin\tsc.ps1:9 char:20
+ ... sedir/pwsh"  "$basedir/../Data/global/node_modules/.bin/tsc.ps1" "$@"
+                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~Unexpected token '"$basedir/../Data/global/node_modules/.bin/tsc.ps1"' in expression or statement.

At C:\Users\chgagnon\AppData\Local\Yarn\bin\tsc.ps1:9 char:72
+ ... sedir/pwsh"  "$basedir/../Data/global/node_modules/.bin/tsc.ps1" "$@"
+                                                                      ~~~~Unexpected token '"$@"' in expression or statement.
   at System.Management.Automation.ScriptBlock.Create(Parser parser, String fileName, String fileContents)
...
```

**Comment by andschwa** [comment]
...is it being run under Cygwin?

> +     *CYGWIN*) basedir=`cygpath -w "$basedir"`;;

**Comment by Charles-Gagnon** [comment]
Not purposely...the callstack is very weird. I have no idea what might be causing this but it goes away if I uninstall your extension...

**Comment by andschwa** [comment]
Uhh @rjmholt any ideas?!

**Comment by rjmholt** [comment]
This looks like something is trying to run a POSIX shell script with PowerShell

**Comment by rjmholt** [comment]
Never seen anything like it before though

**Comment by rjmholt** [comment]
It looks like there are three languages involved here.

First a promise is being rejected, which is a JS/TS concept. So some TS program is running things at the top level. I assume through yarn or trying to run yarn.

Then the script that it's trying to run or parse has a ps1 extension, but wherever it's being handed to PowerShell it's failing to parse. What's more, it's not a normal parse error because it's telling you the API call that failed. It sounds like yarn is calling into something that tries to parse the script as a PowerShell script?

And then the thing being parsed is *actually* Bourne shell script. The syntax that it's failing on is all classic Bourne shell.

**Comment by rjmholt** [comment]
Any idea what is executing this code and when? Maybe a stack trace on the rejected promise?

**Comment by andschwa** [comment]
Is `bin\tsc.ps1` the "TypeScript Compiler (PowerShell script)"? Just checking an assumption here.

**Comment by rjmholt** [comment]
Looks like it's a script generated by [this file](https://github.com/npm/cmd-shim/blob/master/index.js), used by npm.

For some reason the [sh script contents](https://github.com/npm/cmd-shim/blob/c5118da34126e6639361fe9706a5ff07e726ed45/index.js#L131-L151) seem to be in the [ps1 script](https://github.com/npm/cmd-shim/blob/c5118da34126e6639361fe9706a5ff07e726ed45/index.js#L223).

Based on the top-level error message, I think this is [a VSCode execution API](https://github.com/microsoft/vscode/blob/5728f84b8c3f6d8689bd261a0017c990352e420c/src/vs/workbench/services/extensions/node/extensionHostProcessSetup.ts#L245) trying to run that file somehow

**Comment by rjmholt** [comment]
@Charles-Gagnon should I perform a user install of ADS to repro this?

**Comment by rjmholt** [comment]
I'm having trouble reproducing it myself:

image

**Comment by rjmholt** [comment]
My suspicion is that something might be set in your local environment. In particular, the `sh` script has no extension and I wonder if for some reason that's just being picked up to run as a PowerShell script.

I think if we can find more information about where and when this error is occurring on your machine, we might be able to dig deeper.

**Comment by Charles-Gagnon** [comment]
Interesting, especially since the older (current) version of the extension seems to work fine. @alanrenmsft could you try installing the extension to see if it seems to work for you? If it does then we can probably release it as long as @rjmholt and @andschwa have signed off on it.

**Comment by Charles-Gagnon** [comment]
(I'll look more into this later, have some other things I need to take care of first)

**Comment by Charles-Gagnon** [comment]
So this is coming from the languageclient - it's sending a `powershell/getCommand` message which is then causing this error.

I'm poking around in the language service to see if anything sticks out as to why this would be failing like this

**Comment by Charles-Gagnon** [comment]
So the callstack of interest seems to be here : 

```
   at System.Management.Automation.ScriptBlock.Create(Parser parser, String fileName, String fileContents)
   at System.Management.Automation.ExternalScriptInfo.ParseScriptContents(Parser parser, String fileName, String fileContents, Nullable`1 definingLanguageMode)
   at System.Management.Automation.ExternalScriptInfo.get_ScriptBlock()
   at System.Management.Automation.ExternalScriptInfo.get_CommandMetadata()
   at System.Management.Automation.CommandInfo.GenerateCommandParameterSetInfo()
   at System.Management.Automation.CommandInfo.get_ParameterSets()
   at Microsoft.PowerShell.EditorServices.Handlers.GetCommandHandler.<Handle>d__3.MoveNext()
   ```
   
   from https://github.com/PowerShell/PowerShellEditorServices/blob/master/src/PowerShellEditorServices/Services/PowerShellContext/Handlers/GetCommandHandler.cs#L79
   
  I looked around a bit but being not familiar with the code I'm not exactly sure what this System.Management.Automation.CommandInfo stuff is. Are you familiar enough with that to know what it's doing under the covers?
  
  (side note - this actually repros in VS Code too so it does appear to be some environment problem for me)

**Comment by rjmholt** [comment]
> You should be able to get the same output in the Integrated Console by running `gcm | ? { $_.Name -like 'tsc*' }`.

I'd be interested to see if this turns up anything btw.

**Comment by andschwa** [comment]
> In this case, the command is an `ExternalScriptInfo`, meaning PowerShell has found it as a script file (so it's some file with a `.ps1` extension). In trying to collect the parameter set metadata of the command, under the hood the `ExternalScriptInfo` lazily parses the script (i.e. the script isn't parsed until something that needs that information forces the parsing). And when that happens, PowerShell discovers that the script isn't valid PowerShell and [throws](https://github.com/PowerShell/PowerShell/blob/658837323599ab1c7a81fe66fcd43f7420e4402b/src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs#L649-L653).

Do you think we should perhaps _not_ throw in the extension? It doesn't seem to be an error we can do anything about except _not_ have gathered commands for the explorer.

**Comment by Charles-Gagnon** [comment]
That command didn't come up with anything, but I found the scripts and removing them finally unblocked me here. This has been quite the journey! I'm going to open an issue in your repo with the details to follow up with - this ended up being an issue with npm it appears.

**Comment by Charles-Gagnon** [comment]
So at this point I'm good to push the gallery update and get this published - does that work for you two?

**Review by Charles-Gagnon — APPROVED** [review]

**Comment by andschwa** [comment]
> So at this point I'm good to push the gallery update and get this published - does that work for you two?

Good with me but wait for Rob to chime in too.

**Comment by rjmholt** [comment]
> Do you think we should perhaps not throw in the extension? It doesn't seem to be an error we can do anything about except not have gathered commands for the explorer.

Yeah we need to be robust to errors like that. Those particular properties can be quite nasty because they're lazily evaluated. In other forms of `CommandInfo`, evaluating them off the pipeline thread can be treacherous. PSScriptAnalyzer suffers from intermittent `NullReferenceException`s as a result of `AliasInfo` object properties.

> So at this point I'm good to push the gallery update and get this published - does that work for you two?

Yep that works for me! Thanks for following up with us @Charles-Gagnon.

**Comment by Charles-Gagnon** [comment]
Gallery updated!

**Comment by Charles-Gagnon** [comment]
@rjmholt I just noticed that you had already pointed out the script earlier - https://github.com/npm/cmd-shim/blob/master/index.js

I just missed that 🤦‍♂️ But yeah that's the one!

Anyways here's the issue for following up with : https://github.com/PowerShell/vscode-powershell/issues/3324

**Pull request merged**